Find View JavaFX rewrite with extra functionality

Atlas139mkm opened the pull request:
<!--

### Requirements

* Filling out the template is required. Any pull request that does not include
enough information to be reviewed in a timely manner may be closed at the
maintainers' discretion.
* All new code requires unit tests to ensure they work as expected and will
continue to work as new code is added in the future (regression testing).
* Make sure your branch name is prefixed by `feature`, `bugfix` or `release`
* Have you read Constellation's Code of Conduct? By filing an issue, you are
expected to comply with it, including treating everyone with respect:
https://github.com/constellation-app/constellation/blob/master/CODE_OF_CONDUCT.md

-->

### Prerequisites

- [ ] Reviewed the [checklist](CHECKLIST.md)

- [ ] Reviewed feedback from the "Sonar Cloud" bot. Note that you have to wait
    for the "CI / Unit Tests") to complete first.

### Description of the Change

Okay so this is just a draft so I can get something up on github for other people to look at and give me feedback. This is still in rough shape with no tests yet, however I wanted to show what I have done up to this point.

- There will probably be a lot of:
   - code smells I need to fix
   - finals I need to add
   - if statement optimisation I can do
   - Breaking up of methods to increase testability

- My current progress
   - Created UI for new find view
   - Fully functional basic find (A few minor things I still need to implement)
   - User can basic find on all open graphs
   - User can cycle results by clicking find next and find prev
   - I will post a gif below of me showing off the usability 

- My current next steps (no particular order)
   - Implement replace
   - fix issues surrounding user deleting graph elements that have been found (I have a solution in mind, just no implemented)
   - auto focus find text field
   - ctrl f to open new find view
   - Write tests
   - Default Find View to float in the top right of the users screen
   - fix IO issues when saving
   - Basic find in currently selected subset

- The code contains:
   - Top component
   - UI pane, UI tabs
   - Controller (singleton) - manages the interaction between the UI and the graph
   - BasicFindReplaceParameters - a object that contains all the UI elements in logic form
   - BasicFindPlugin - a remastered version of the old plugin that allows find next / prev
   - FindViewConcept and FindViewStateAttributeDescription - Used for a meta attribute that holds a list of results found
   - FindResult - A result that contains the element type, relevant IDs that allow the logic to find it on the graph

Gif 1 - Selecting all with label or identifier == vertex cycling through next and prev
![FindNextPrev](https://user-images.githubusercontent.com/88649439/137060564-79ac4e09-b02d-4171-a5fd-8d8f5770787a.gif)

Gif 2 - Searching Adelaide on all open graphs - selects Adelaide on both
![SearchMutlipleGraphs](https://user-images.githubusercontent.com/88649439/137060827-4f2889d4-6029-4ebb-a18e-79a40b580c1a.gif)

- Gifs might not show it well, but the speed is very fast. Since it is done by the existing search algorithm it is on par with that.

These are the two major extra features that I have configured the find view to be capable of. Please ask away any questions people have, and please give any feedback if you think what I have needs serious improvement or is going in the right direction.


### Applicable Issues

#55 
#856
<!-- Link any applicable issues here -->

arcturus2 commented:
@Atlas139mkm very nice :+1:. I think you are on the right track and I am keen to see what you do with the "Advance Search".

To start with can you please restructure the packages so they closely align with the "Core Data Access View" and "Core Table View" modules? So rather than use the `gui` package call it `components` and the `FindViewConcept` should be in the `state` package etc. Until now we have not really had a structure for views but those 2 views are forming the formal approach which we can start to document soon.

Performance is fine and works really well across multiple graphs so I won't bother profiling it yet.

Atlas139mkm commented:
@arcturus2 yep can do, I forgot to mention I have only loosely done the packages. I'll use the ones you suggested as examples.

Atlas139mkm commented:
Just pushing changes as I do them

Atlas139mkm commented:
UPDATE:

- Find view can now open as default float in the top right. I have created a note comment in the code about this solution. It is not flawless as it works of the basis that no other views are set to mode float however I could not figure out another way. For the moment this works well.
![openAsFloat](https://user-images.githubusercontent.com/88649439/139605836-c2b271a7-2ac2-440a-ab5b-962b7abcf1d0.gif)

Current Project status:
- View now defaults as float
- Fixed JX issue with listed
- Now auto selects ignore option
- Added Unit tests to all major classes (Basic find tab, basic replace tab, find view controller, find view top component, basic find plugin, replace plugin and reset state plugin)

remaining tasks
- attempt to fix up as many code smells as possible

remaining issues:
- auto focus on text box still does not work on the FIRST occurrence of find view opening.

Atlas139mkm commented:
Alright, the view has no known issues. Other than: 
auto focus on text box still does not work on the FIRST occurrence of find view opening. (even this is inconsistent however)

I'm requesting a PR review now as it should be mostly small touch up issues I need to fix. Let me know your feedback so we can try and get this in experimental for user feedback ASAP.

antares1470 commented on CoreFindView/src/au/gov/asd/tac/constellation/views/find2/components/ReplaceTab.java:
```diff
@@ -0,0 +1,178 @@
+
+        // determine what currentSelectionChoiceBox option is selected
+        switch (currentSelectionChoiceBox.getSelectionModel().getSelectedIndex()) {
+            case 0:
+                break;
```
given this is the same as the default case, I would either omit this altogether or shift it below `case 1` if you do want to explicilty mention it. i.e.:
```
...
case 0:
default:
  break;
```

antares1470 commented on CoreFindView/src/au/gov/asd/tac/constellation/views/find2/plugins/BasicFindPlugin.java:
```diff
@@ -0,0 +1,332 @@
+                        // Determine if the find string matches the attribute 
+                        // string
+                        Matcher match = searchPattern.matcher(value);
+                        if (matchWholeWord) {
+                            found = match.matches();
+                        } else {
+                            found = match.find();
+                        }
```
`found = matchWholeWord ? match.matches() : match.find()`

antares1470 commented on CoreFindView/src/au/gov/asd/tac/constellation/views/find2/plugins/BasicFindPlugin.java:
```diff
@@ -0,0 +1,332 @@
+
+        }
+        //If no results are found, set the meta attribute to null
+        if (foundResult.isEmpty()) {
+            graph.setObjectValue(stateId, 0, null);
+        } else {
+            graph.setObjectValue(stateId, 0, foundResult);
+        }
```
`graph.setObjectValue(stateId, 0, foundResult.isEmpty() ? null : foundResult);`

antares1470 commented on CoreFindView/src/au/gov/asd/tac/constellation/views/find2/plugins/BasicFindPlugin.java:
```diff
@@ -0,0 +1,332 @@
+     * @param graph
+     * @param selectedAttribute
+     */
+    private void selectFindInResults(boolean findInCurrentSelection, FindResultsList findInCurrentSelectionList,
```
parameters can be final

antares1470 commented on CoreFindView/src/au/gov/asd/tac/constellation/views/find2/plugins/BasicFindPlugin.java:
```diff
@@ -0,0 +1,332 @@
+     * @param graph
+     * @param selectedAttribute
+     */
+    private void selectRemoveFromResults(boolean removeFromCurrentSelection, FindResultsList removeFromCurrentSelectionList,
```
same here

antares1470 commented on CoreFindView/src/au/gov/asd/tac/constellation/views/find2/plugins/ResetStatePlugin.java:
```diff
@@ -0,0 +1,74 @@
+     * @throws PluginException
+     */
+    @Override
+    protected void edit(GraphWriteMethods graph, PluginInteraction interaction, PluginParameters parameters) throws InterruptedException, PluginException {
```
parameters final

antares1470 commented on CoreFindView/src/au/gov/asd/tac/constellation/views/find2/plugins/ResetStatePlugin.java:
```diff
@@ -0,0 +1,74 @@
+     */
+    @Override
+    protected void edit(GraphWriteMethods graph, PluginInteraction interaction, PluginParameters parameters) throws InterruptedException, PluginException {
+        int stateId = FindViewConcept.MetaAttribute.FINDVIEW_STATE.ensure(graph);
+        FindResultsList foundResult = graph.getObjectValue(stateId, 0);
```
finals

antares1470 commented on CoreFindView/src/au/gov/asd/tac/constellation/views/find2/state/FindViewStateIoProvider.java:
```diff
@@ -0,0 +1,167 @@
+    }
+
+    @Override
+    public void readObject(int attributeId, int elementId, JsonNode jnode, GraphWriteMethods writableGraph, Map<Integer, Integer> vertexMap, Map<Integer, Integer> transactionMap, GraphByteReader byteReader, ImmutableObjectCache cache) throws IOException {
```
parameters final

antares1470 commented on CoreFindView/src/au/gov/asd/tac/constellation/views/find2/state/FindViewStateIoProvider.java:
```diff
@@ -0,0 +1,167 @@
+            final List<FindResult> findResults = new ArrayList<>();
+            final ArrayNode findResultsArray = (ArrayNode) jnode.withArray("findResults");
+            for (int i = 0; i < findResultsArray.size(); i = i + 3) {
+                if (findResultsArray.get(i).isNull()) {
+                    findResults.add(null);
+                } else {
+                    findResults.add(new FindResult(findResultsArray.get(i).asInt(),
+                            findResultsArray.get(i + 1).asInt(), GraphElementType.getValue(findResultsArray.get(i + 2).asText())));
+                }
```
`findResults.add(findResultsArray.get(i).isNull() ? null : new FindResult(...))`

antares1470 commented on CoreFindView/src/au/gov/asd/tac/constellation/views/find2/state/FindViewStateIoProvider.java:
```diff
@@ -0,0 +1,167 @@
+    }
+
+    @Override
+    public void writeObject(Attribute attribute, int elementId, JsonGenerator jsonGenerator,
```
parameters final

antares1470 commented on CoreFindView/src/au/gov/asd/tac/constellation/views/find2/state/FindViewStateIoProvider.java:
```diff
@@ -0,0 +1,167 @@
+
+                // Stores a list of the Find Results, ID, UID and Graph element type label
+                jsonGenerator.writeArrayFieldStart("findResults");
+                for (FindResult fr : resultsList) {
```
final

antares1470 commented on CoreFindView/src/au/gov/asd/tac/constellation/views/find2/state/FindViewStateIoProvider.java:
```diff
@@ -0,0 +1,167 @@
+                jsonGenerator.writeEndArray();
+
+                // Store all the basic find replace parameters
+                BasicFindReplaceParameters parameters = resultsList.getSearchParameters();
```
final

antares1470 commented on CoreFindView/src/au/gov/asd/tac/constellation/views/find2/state/FindViewStateIoProvider.java:
```diff
@@ -0,0 +1,167 @@
+
+                // Store all the selected attributes of the search
+                jsonGenerator.writeArrayFieldStart("selectedAttributes");
+                for (Attribute a : parameters.getAttributeList()) {
```
final

antares1470 commented on CoreFindView/src/au/gov/asd/tac/constellation/views/find2/utilities/BasicFindReplaceParameters.java:
```diff
@@ -0,0 +1,247 @@
+    @Override
+    public boolean equals(final Object object) {
+        int matches = 0;
+        if (object == null) {
+            return false;
+        }
+        if (getClass() != object.getClass()) {
```
merge these ifs together (with `||`)

antares1470 commented on CoreFindView/src/au/gov/asd/tac/constellation/views/find2/utilities/FindResult.java:
```diff
@@ -0,0 +1,200 @@
+    }
+
+    @Override
+    public boolean equals(Object obj) {
```
parameter final

antares1470 commented on CoreFindView/src/au/gov/asd/tac/constellation/views/find2/utilities/FindResult.java:
```diff
@@ -0,0 +1,200 @@
+        if (!(obj instanceof FindResult)) {
+            return false;
+        }
+        FindResult other = (FindResult) obj;
```
final

antares1470 commented on CoreFindView/src/au/gov/asd/tac/constellation/views/find2/utilities/FindResultsList.java:
```diff
@@ -0,0 +1,125 @@
+     *
+     * @param resultsList
+     */
+    public FindResultsList(FindResultsList resultsList) {
```
parameter final

Atlas139mkm commented on CoreFindView/src/au/gov/asd/tac/constellation/views/find/gui/FindTopComponent.java:
```diff
@@ -84,7 +84,7 @@
 @ActionReferences({
     @ActionReference(path = "Menu/Edit", position = 1000, separatorBefore = 999),
-    @ActionReference(path = "Shortcuts", name = "C-F")
+    @ActionReference(path = "Shortcuts", name = "C-B")
```
I believe I did this for my own testing sake. I have changed find2 to ctrl b and changed find back to ctrl f

Atlas139mkm commented on CoreFindView/src/au/gov/asd/tac/constellation/views/find2/components/BasicFindTab.java:
```diff
@@ -0,0 +1,605 @@
+
+        // set the action for changing the seleciton in the
+        // currentSelctionChoiceBox
+        currentSelectionChoiceBox.getSelectionModel().selectedItemProperty().addListener(new ChangeListener<String>() {
```
Converting this one to a lambda was having difficulties. 

currentSelectionChoiceBox.getSelectionModel().selectedItemProperty().addListener((ObservableValue<? extends String> observableValue, String oldElement, String newElement) -> {
            updateSelectionFactors();
        });

This was not working so, ill leave it for now, unless you have another suggest
ion of how to write it.

aldebaran30701 commented:
Just looking at the functionality, I wanted to bring up a few discussion points just to entertain other ideas around them.
1. Button ordering [find prev, find next, find all]
I feel like the furthermost left item would be better suited as the most common options (find all). I also feel like having the whole word previous instead of prev would better suit the UI even though the buttons would be sized differently.

2. RegEx changed to RegularExpression for easier recognition.
This might help a user understand what RegEx is or allow it to be searched easier

Atlas139mkm commented:
![image](https://user-images.githubusercontent.com/88649439/141239473-5c90c5a8-b913-4768-b06a-f384466d338d.png)
 
 Thoughts?

antares1470 commented:
>     2\. RegEx changed to RegularExpression for easier recognition.
>         This might help a user understand what RegEx is or allow it to be searched easier

@aldebaran30701 it's probably worth noting that "RegEx" is what is done in the old Find View (that's not to say that makes it a better option, but rather I assume this was then just copied across). For me though, given I often hear regular expression referred to a regex, I would guess that those who don't understand what "RegEx" is are probably not going to have a much better time understanding what "RegularExpression" is

antares1470 commented on CoreFindView/src/au/gov/asd/tac/constellation/views/find2/FindViewController.java:
```diff
@@ -107,10 +109,10 @@ public FindViewController init(final FindViewTopComponent parentComponent) {
                 future.get();
             } catch (final InterruptedException ex) {
-                Exceptions.printStackTrace(ex);
+                LOGGER.log(Level.SEVERE, ex.getMessage());
```
These ones where you've replaced the stack trace print with a log (similar in other classes) are you able to add the exception as a parameter?
(e.g. `LOGGER.log(Level.SEVERE, ex.getMessage(), ex)`)

antares1470 commented on CoreFindView/test/unit/src/au/gov/asd/tac/constellation/views/find2/components/FindViewTabsNGTest.java:
```diff
@@ -77,6 +77,26 @@ public void setUpMethod() throws Exception {
+     * Test listener for changing selected tab
+     */
+    @Test
+    public void testChangeTab() {
```
Was there something wrong with this test? Just noticing all the code in it is commented out

Atlas139mkm replied:
Yeah I believe it just wasn't working originally then I forgot about it.

antares1470 commented:
@Atlas139mkm I've done some testing of your PR as well and what I've tested so far looks good (of the basic find and replace)

Atlas139mkm commented:
*IMPORTANT* I have commented out the implementation of auto save for the sake of the CI tests passing. Before this is ever merged in uncomment that code or look into the CI tests failing due to auto save triggering. The file is "autosavestartup.java" and the commented out lines are 58-120.

Atlas139mkm commented:
<p class=MsoNormal><b><u>Find View notes:<o:p></o:p></u></b></p>
<p class=MsoNormal>The basic find, replace and advanced find are all functionally
complete. There is a few tests that still need to be written but nothing should
come up that drastically effects the main code base. I have done a fair amount
of my own user testing to ensure that a lot of different scenarios play out as intended.
I have written a list of various things to note regarding sonar cloud, unit
tests, known minor bugs and the CI failing from autosave. <o:p></o:p></p>

<p class=MsoNormal><b><u><o:p><span style='text-decoration:none'>&nbsp;</span></o:p></u></b></p>

<p class=MsoNormal><b>Sonar cloud<o:p></o:p></b></p>
<p class=MsoNormal>Current Standing code coverage is 60.5%. This number is false
due to an issue with several extra classes added to the branch. This is explained
below. The true number is probably closer to 65-75%<o:p></o:p></p>

<p class=MsoNormal>There are 39 code smells. 8 of these are related to the extra
classes explained below. Most of the remaining code smells are comment
coverage. Each class should have comments that appropriately explain the functionality
of how each aspect works.<o:p></o:p></p>

<p class=MsoNormal>In sonar cloud it shows the list of classes with new code.
Currently 12 of these classes were not modified by me. I believe this was from
me messing up pulling changes from master. This currently lowers the % of code
coverage. These classes can be ignored:<o:p></o:p></p>

<p class=MsoListParagraphCxSpFirst style='text-indent:-18.0pt;mso-list:l0 level1 lfo2'><![if !supportLists]><span
style='mso-ascii-font-family:Calibri;mso-fareast-font-family:Calibri;
mso-hansi-font-family:Calibri;mso-bidi-font-family:Calibri'><span
style='mso-list:Ignore'>-<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span></span><![endif]><a
href="https://sonarcloud.io/component_measures?id=constellation-app_constellation&amp;metric=new_coverage&amp;pullRequest=1489&amp;selected=constellation-app_constellation%3ACoreVisualSchema%2Fsrc%2Fau%2Fgov%2Fasd%2Ftac%2Fconstellation%2Fgraph%2Fschema%2Fvisual%2Fattribute%2FCameraAttributeDescription.java"><span
style='font-size:10.0pt;line-height:107%;font-family:"Helvetica",sans-serif;
color:black;mso-color-alt:windowtext;background:white;text-decoration:none;
text-underline:none'><span
title="constellation-app_constellation:CoreVisualSchema/src/au/gov/asd/tac/constellation/graph/schema/visual/attribute/CameraAttributeDescription.java">CameraAttributeDescription.java</span></span></a><o:p></o:p></p>

<p class=MsoListParagraphCxSpMiddle style='text-indent:-18.0pt;mso-list:l0 level1 lfo2'><![if !supportLists]><span
style='mso-ascii-font-family:Calibri;mso-fareast-font-family:Calibri;
mso-hansi-font-family:Calibri;mso-bidi-font-family:Calibri'><span
style='mso-list:Ignore'>-<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span></span><![endif]><a
href="https://sonarcloud.io/component_measures?id=constellation-app_constellation&amp;metric=new_coverage&amp;pullRequest=1489&amp;selected=constellation-app_constellation%3ACoreVisualSchema%2Fsrc%2Fau%2Fgov%2Fasd%2Ftac%2Fconstellation%2Fgraph%2Fschema%2Fvisual%2Fattribute%2FConnectionModeAttributeDescription.java"><span
style='font-size:10.0pt;line-height:107%;font-family:"Helvetica",sans-serif;
color:black;mso-color-alt:windowtext;background:whitesmoke;text-decoration:
none;text-underline:none'><span
title="constellation-app_constellation:CoreVisualSchema/src/au/gov/asd/tac/constellation/graph/schema/visual/attribute/ConnectionModeAttributeDescription.java">ConnectionModeAttributeDescription.java</span></span></a><o:p></o:p></p>

<p class=MsoListParagraphCxSpMiddle style='text-indent:-18.0pt;mso-list:l0 level1 lfo2'><![if !supportLists]><span
style='mso-ascii-font-family:Calibri;mso-fareast-font-family:Calibri;
mso-hansi-font-family:Calibri;mso-bidi-font-family:Calibri'><span
style='mso-list:Ignore'>-<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span></span><![endif]><a
href="https://sonarcloud.io/component_measures?id=constellation-app_constellation&amp;metric=new_coverage&amp;pullRequest=1489&amp;selected=constellation-app_constellation%3ACoreVisualSchema%2Fsrc%2Fau%2Fgov%2Fasd%2Ftac%2Fconstellation%2Fgraph%2Fschema%2Fvisual%2Fattribute%2FDecoratorsAttributeDescription.java"><span
style='font-size:10.0pt;line-height:107%;font-family:"Helvetica",sans-serif;
color:black;mso-color-alt:windowtext;background:white;text-decoration:none;
text-underline:none'><span
title="constellation-app_constellation:CoreVisualSchema/src/au/gov/asd/tac/constellation/graph/schema/visual/attribute/DecoratorsAttributeDescription.java">DecoratorsAttributeDescription.java</span></span></a><o:p></o:p></p>

<p class=MsoListParagraphCxSpMiddle style='text-indent:-18.0pt;mso-list:l0 level1 lfo2'><![if !supportLists]><span
style='mso-ascii-font-family:Calibri;mso-fareast-font-family:Calibri;
mso-hansi-font-family:Calibri;mso-bidi-font-family:Calibri'><span
style='mso-list:Ignore'>-<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span></span><![endif]><a
href="https://sonarcloud.io/component_measures?id=constellation-app_constellation&amp;metric=new_coverage&amp;pullRequest=1489&amp;selected=constellation-app_constellation%3ACoreVisualSchema%2Fsrc%2Fau%2Fgov%2Fasd%2Ftac%2Fconstellation%2Fgraph%2Fschema%2Fvisual%2Fattribute%2FDrawFlagsAttributeDescription.java"><span
style='font-size:10.0pt;line-height:107%;font-family:"Helvetica",sans-serif;
color:black;mso-color-alt:windowtext;background:whitesmoke;text-decoration:
none;text-underline:none'><span
title="constellation-app_constellation:CoreVisualSchema/src/au/gov/asd/tac/constellation/graph/schema/visual/attribute/DrawFlagsAttributeDescription.java">DrawFlagsAttributeDescription.java</span></span></a><o:p></o:p></p>

<p class=MsoListParagraphCxSpMiddle style='text-indent:-18.0pt;mso-list:l0 level1 lfo2'><![if !supportLists]><span
style='mso-ascii-font-family:Calibri;mso-fareast-font-family:Calibri;
mso-hansi-font-family:Calibri;mso-bidi-font-family:Calibri'><span
style='mso-list:Ignore'>-<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span></span><![endif]><a
href="https://sonarcloud.io/component_measures?id=constellation-app_constellation&amp;metric=new_coverage&amp;pullRequest=1489&amp;selected=constellation-app_constellation%3ACoreUtilities%2Fsrc%2Fau%2Fgov%2Fasd%2Ftac%2Fconstellation%2Futilities%2Fimage%2FGaussianBlur.java"><span
style='font-size:10.0pt;line-height:107%;font-family:"Helvetica",sans-serif;
color:black;mso-color-alt:windowtext;background:whitesmoke;text-decoration:
none;text-underline:none'><span
title="constellation-app_constellation:CoreUtilities/src/au/gov/asd/tac/constellation/utilities/image/GaussianBlur.java">GaussianBlur.java</span></span></a><o:p></o:p></p>

<p class=MsoListParagraphCxSpMiddle style='text-indent:-18.0pt;mso-list:l0 level1 lfo2'><![if !supportLists]><span
style='mso-ascii-font-family:Calibri;mso-fareast-font-family:Calibri;
mso-hansi-font-family:Calibri;mso-bidi-font-family:Calibri'><span
style='mso-list:Ignore'>-<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span></span><![endif]><a
href="https://sonarcloud.io/component_measures?id=constellation-app_constellation&amp;metric=new_coverage&amp;pullRequest=1489&amp;selected=constellation-app_constellation%3ACoreHistogramView%2Fsrc%2Fau%2Fgov%2Fasd%2Ftac%2Fconstellation%2Fviews%2Fhistogram%2FHistogramDisplay.java"><span
style='font-size:10.0pt;line-height:107%;font-family:"Helvetica",sans-serif;
color:black;mso-color-alt:windowtext;background:white;text-decoration:none;
text-underline:none'><span
title="constellation-app_constellation:CoreHistogramView/src/au/gov/asd/tac/constellation/views/histogram/HistogramDisplay.java">HistogramDisplay.java</span></span></a><o:p></o:p></p>

<p class=MsoListParagraphCxSpMiddle style='text-indent:-18.0pt;mso-list:l0 level1 lfo2'><![if !supportLists]><span
style='mso-ascii-font-family:Calibri;mso-fareast-font-family:Calibri;
mso-hansi-font-family:Calibri;mso-bidi-font-family:Calibri'><span
style='mso-list:Ignore'>-<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span></span><![endif]><a
href="https://sonarcloud.io/component_measures?id=constellation-app_constellation&amp;metric=new_coverage&amp;pullRequest=1489&amp;selected=constellation-app_constellation%3ACoreVisualSchema%2Fsrc%2Fau%2Fgov%2Fasd%2Ftac%2Fconstellation%2Fgraph%2Fschema%2Fvisual%2Fattribute%2FLineStyleAttributeDescription.java"><span
style='font-size:10.0pt;line-height:107%;font-family:"Helvetica",sans-serif;
color:black;mso-color-alt:windowtext;background:white;text-decoration:none;
text-underline:none'><span
title="constellation-app_constellation:CoreVisualSchema/src/au/gov/asd/tac/constellation/graph/schema/visual/attribute/LineStyleAttributeDescription.java">LineStyleAttributeDescription.java</span></span></a><o:p></o:p></p>

<p class=MsoListParagraphCxSpMiddle style='text-indent:-18.0pt;mso-list:l0 level1 lfo2'><![if !supportLists]><span
style='mso-ascii-font-family:Calibri;mso-fareast-font-family:Calibri;
mso-hansi-font-family:Calibri;mso-bidi-font-family:Calibri'><span
style='mso-list:Ignore'>-<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span></span><![endif]><a
href="https://sonarcloud.io/component_measures?id=constellation-app_constellation&amp;metric=new_coverage&amp;pullRequest=1489&amp;selected=constellation-app_constellation%3ACoreMapView%2Fsrc%2Fau%2Fgov%2Fasd%2Ftac%2Fconstellation%2Fviews%2Fmapview%2FMapViewTopComponent.java"><span
style='font-size:10.0pt;line-height:107%;font-family:"Helvetica",sans-serif;
color:black;mso-color-alt:windowtext;background:whitesmoke;text-decoration:
none;text-underline:none'><span
title="constellation-app_constellation:CoreMapView/src/au/gov/asd/tac/constellation/views/mapview/MapViewTopComponent.java">MapViewTopComponent.java</span></span></a><o:p></o:p></p>

<p class=MsoListParagraphCxSpMiddle style='text-indent:-18.0pt;mso-list:l0 level1 lfo2'><![if !supportLists]><span
style='mso-ascii-font-family:Calibri;mso-fareast-font-family:Calibri;
mso-hansi-font-family:Calibri;mso-bidi-font-family:Calibri'><span
style='mso-list:Ignore'>-<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span></span><![endif]><a
href="https://sonarcloud.io/component_measures?id=constellation-app_constellation&amp;metric=new_coverage&amp;pullRequest=1489&amp;selected=constellation-app_constellation%3ACoreUtilities%2Fsrc%2Fau%2Fgov%2Fasd%2Ftac%2Fconstellation%2Futilities%2Frest%2FResponse.java"><span
style='font-size:10.0pt;line-height:107%;font-family:"Helvetica",sans-serif;
color:black;mso-color-alt:windowtext;background:white;text-decoration:none;
text-underline:none'><span
title="constellation-app_constellation:CoreUtilities/src/au/gov/asd/tac/constellation/utilities/rest/Response.java">Response.java</span></span></a><o:p></o:p></p>

<p class=MsoListParagraphCxSpMiddle style='text-indent:-18.0pt;mso-list:l0 level1 lfo2'><![if !supportLists]><span
style='mso-ascii-font-family:Calibri;mso-fareast-font-family:Calibri;
mso-hansi-font-family:Calibri;mso-bidi-font-family:Calibri'><span
style='mso-list:Ignore'>-<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span></span><![endif]><a
href="https://sonarcloud.io/component_measures?id=constellation-app_constellation&amp;metric=new_coverage&amp;pullRequest=1489&amp;selected=constellation-app_constellation%3ACoreVisualSchema%2Fsrc%2Fau%2Fgov%2Fasd%2Ftac%2Fconstellation%2Fgraph%2Fschema%2Fvisual%2Fattribute%2FTransactionGraphLabelsAttributeDescription.java"><span
style='font-size:10.0pt;line-height:107%;font-family:"Helvetica",sans-serif;
color:black;mso-color-alt:windowtext;background:whitesmoke;text-decoration:
none;text-underline:none'><span
title="constellation-app_constellation:CoreVisualSchema/src/au/gov/asd/tac/constellation/graph/schema/visual/attribute/TransactionGraphLabelsAttributeDescription.java">TransactionGraphLabelsAttributeDescription.java</span></span></a><o:p></o:p></p>

<p class=MsoListParagraphCxSpMiddle style='text-indent:-18.0pt;mso-list:l0 level1 lfo2'><![if !supportLists]><span
style='mso-ascii-font-family:Calibri;mso-fareast-font-family:Calibri;
mso-hansi-font-family:Calibri;mso-bidi-font-family:Calibri'><span
style='mso-list:Ignore'>-<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span></span><![endif]><a
href="https://sonarcloud.io/component_measures?id=constellation-app_constellation&amp;metric=new_coverage&amp;pullRequest=1489&amp;selected=constellation-app_constellation%3ACoreVisualSchema%2Fsrc%2Fau%2Fgov%2Fasd%2Ftac%2Fconstellation%2Fgraph%2Fschema%2Fvisual%2Fattribute%2FVertexGraphLabelsAttributeDescription.java"><span
style='font-size:10.0pt;line-height:107%;font-family:"Helvetica",sans-serif;
color:black;mso-color-alt:windowtext;background:whitesmoke;text-decoration:
none;text-underline:none'><span
title="constellation-app_constellation:CoreVisualSchema/src/au/gov/asd/tac/constellation/graph/schema/visual/attribute/VertexGraphLabelsAttributeDescription.java">VertexGraphLabelsAttributeDescription.java</span></span></a><o:p></o:p></p>

<p class=MsoListParagraphCxSpLast style='text-indent:-18.0pt;mso-list:l0 level1 lfo2'><![if !supportLists]><span
style='mso-ascii-font-family:Calibri;mso-fareast-font-family:Calibri;
mso-hansi-font-family:Calibri;mso-bidi-font-family:Calibri'><span
style='mso-list:Ignore'>-<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span></span><![endif]><a
href="https://sonarcloud.io/component_measures?id=constellation-app_constellation&amp;metric=new_coverage&amp;pullRequest=1489&amp;selected=constellation-app_constellation%3ACoreVisualSchema%2Fsrc%2Fau%2Fgov%2Fasd%2Ftac%2Fconstellation%2Fgraph%2Fschema%2Fvisual%2Fattribute%2FBlazeAttributeDescription.java"><span
style='font-size:10.0pt;line-height:107%;font-family:"Helvetica",sans-serif;
color:black;mso-color-alt:windowtext;background:#ECF6FE;text-decoration:none;
text-underline:none'><span
title="constellation-app_constellation:CoreVisualSchema/src/au/gov/asd/tac/constellation/graph/schema/visual/attribute/BlazeAttributeDescription.java">BlazeAttributeDescription.java</span></span></a><o:p></o:p></p>

<p class=MsoNormal><o:p>&nbsp;</o:p></p>

<p class=MsoNormal><b>Classes currently left untested<o:p></o:p></b></p>

<p class=MsoListParagraphCxSpFirst style='text-indent:-18.0pt;mso-list:l0 level1 lfo2'><![if !supportLists]><span
style='mso-ascii-font-family:Calibri;mso-fareast-font-family:Calibri;
mso-hansi-font-family:Calibri;mso-bidi-font-family:Calibri;mso-bidi-font-weight:
bold'><span style='mso-list:Ignore'>-<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span></span><![endif]><a
href="https://sonarcloud.io/component_measures?id=constellation-app_constellation&amp;metric=new_coverage&amp;pullRequest=1489&amp;selected=constellation-app_constellation%3ACoreFindView%2Fsrc%2Fau%2Fgov%2Fasd%2Ftac%2Fconstellation%2Fviews%2Ffind2%2Fcomponents%2Fadvanced%2FDateTimeCriteriaPanel.java"><span
style='font-size:10.0pt;line-height:107%;font-family:"Helvetica",sans-serif;
color:black;mso-color-alt:windowtext;background:white;text-decoration:none;
text-underline:none'><span
title="constellation-app_constellation:CoreFindView/src/au/gov/asd/tac/constellation/views/find2/components/advanced/DateTimeCriteriaPanel.java">DateTimeCriteriaPanel.java</span></span></a><b><o:p></o:p></b></p>

<p class=MsoListParagraphCxSpMiddle style='text-indent:-18.0pt;mso-list:l0 level1 lfo2'><![if !supportLists]><span
style='mso-ascii-font-family:Calibri;mso-fareast-font-family:Calibri;
mso-hansi-font-family:Calibri;mso-bidi-font-family:Calibri;mso-bidi-font-weight:
bold'><span style='mso-list:Ignore'>-<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span></span><![endif]><a
href="https://sonarcloud.io/component_measures?id=constellation-app_constellation&amp;metric=new_coverage&amp;pullRequest=1489&amp;selected=constellation-app_constellation%3ACoreFindView%2Fsrc%2Fau%2Fgov%2Fasd%2Ftac%2Fconstellation%2Fviews%2Ffind2%2Fcomponents%2Fadvanced%2Futilities%2FDateTimeSelector.java"><span
style='font-size:10.0pt;line-height:107%;font-family:"Helvetica",sans-serif;
color:black;mso-color-alt:windowtext;background:#ECF6FE;text-decoration:none;
text-underline:none'><span
title="constellation-app_constellation:CoreFindView/src/au/gov/asd/tac/constellation/views/find2/components/advanced/utilities/DateTimeSelector.java">DateTimeSelector.java</span></span></a><b><o:p></o:p></b></p>

<p class=MsoListParagraphCxSpMiddle style='text-indent:-18.0pt;mso-list:l0 level1 lfo2'><![if !supportLists]><span
style='mso-ascii-font-family:Calibri;mso-fareast-font-family:Calibri;
mso-hansi-font-family:Calibri;mso-bidi-font-family:Calibri;mso-bidi-font-weight:
bold'><span style='mso-list:Ignore'>-<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span></span><![endif]><a
href="https://sonarcloud.io/component_measures?id=constellation-app_constellation&amp;metric=new_coverage&amp;pullRequest=1489&amp;selected=constellation-app_constellation%3ACoreFindView%2Fsrc%2Fau%2Fgov%2Fasd%2Ftac%2Fconstellation%2Fviews%2Ffind2%2Fstate%2FFindViewStateIoProvider.java"><span
style='font-size:10.0pt;line-height:107%;font-family:"Helvetica",sans-serif;
color:black;mso-color-alt:windowtext;background:#ECF6FE;text-decoration:none;
text-underline:none'><span
title="constellation-app_constellation:CoreFindView/src/au/gov/asd/tac/constellation/views/find2/state/FindViewStateIoProvider.java">FindViewStateIoProvider.java</span></span></a><b><o:p></o:p></b></p>

<p class=MsoListParagraphCxSpMiddle style='text-indent:-18.0pt;mso-list:l0 level1 lfo2'><![if !supportLists]><span
style='mso-ascii-font-family:Calibri;mso-fareast-font-family:Calibri;
mso-hansi-font-family:Calibri;mso-bidi-font-family:Calibri;mso-bidi-font-weight:
bold'><span style='mso-list:Ignore'>-<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span></span><![endif]><a
href="https://sonarcloud.io/component_measures?id=constellation-app_constellation&amp;metric=new_coverage&amp;pullRequest=1489&amp;selected=constellation-app_constellation%3ACoreFindView%2Fsrc%2Fau%2Fgov%2Fasd%2Ftac%2Fconstellation%2Fviews%2Ffind2%2Fcomponents%2Fadvanced%2Futilities%2FIconNode.java"><span
style='font-size:10.0pt;line-height:107%;font-family:"Helvetica",sans-serif;
color:black;mso-color-alt:windowtext;background:white;text-decoration:none;
text-underline:none'><span
title="constellation-app_constellation:CoreFindView/src/au/gov/asd/tac/constellation/views/find2/components/advanced/utilities/IconNode.java">IconNode.java</span></span></a><b><o:p></o:p></b></p>

<p class=MsoListParagraphCxSpMiddle style='text-indent:-18.0pt;mso-list:l0 level1 lfo2'><![if !supportLists]><span
style='mso-ascii-font-family:Calibri;mso-fareast-font-family:Calibri;
mso-hansi-font-family:Calibri;mso-bidi-font-family:Calibri;mso-bidi-font-weight:
bold'><span style='mso-list:Ignore'>-<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span></span><![endif]><a
href="https://sonarcloud.io/component_measures?id=constellation-app_constellation&amp;metric=new_coverage&amp;pullRequest=1489&amp;selected=constellation-app_constellation%3ACoreFindView%2Fsrc%2Fau%2Fgov%2Fasd%2Ftac%2Fconstellation%2Fviews%2Ffind2%2Fcomponents%2Fadvanced%2Futilities%2FIconSelector.java"><span
style='font-size:10.0pt;line-height:107%;font-family:"Helvetica",sans-serif;
color:black;mso-color-alt:windowtext;background:whitesmoke;text-decoration:
none;text-underline:none'><span
title="constellation-app_constellation:CoreFindView/src/au/gov/asd/tac/constellation/views/find2/components/advanced/utilities/IconSelector.java">IconSelector.java</span></span></a><b><o:p></o:p></b></p>

<p class=MsoListParagraphCxSpLast style='text-indent:-18.0pt;mso-list:l0 level1 lfo2'><![if !supportLists]><span
style='mso-ascii-font-family:Calibri;mso-fareast-font-family:Calibri;
mso-hansi-font-family:Calibri;mso-bidi-font-family:Calibri;mso-bidi-font-weight:
bold'><span style='mso-list:Ignore'>-<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span></span><![endif]><a
href="https://sonarcloud.io/component_measures?id=constellation-app_constellation&amp;metric=new_coverage&amp;pullRequest=1489&amp;selected=constellation-app_constellation%3ACoreFindView%2Fsrc%2Fau%2Fgov%2Fasd%2Ftac%2Fconstellation%2Fviews%2Ffind2%2Fcomponents%2Fadvanced%2Futilities%2FUseListInputWindow.java"><span
style='font-size:10.0pt;line-height:107%;font-family:"Helvetica",sans-serif;
color:black;mso-color-alt:windowtext;background:white;text-decoration:none;
text-underline:none'><span
title="constellation-app_constellation:CoreFindView/src/au/gov/asd/tac/constellation/views/find2/components/advanced/utilities/UseListInputWindow.java">UseListInputWindow.java</span></span></a><b><o:p></o:p></b></p>

<p class=MsoNormal><o:p>&nbsp;</o:p></p>

<p class=MsoNormal><b>Known issues / quirks<o:p></o:p></b></p>

<p class=MsoListParagraphCxSpFirst style='text-indent:-18.0pt;mso-list:l1 level1 lfo1'><![if !supportLists]><span
style='mso-ascii-font-family:Calibri;mso-fareast-font-family:Calibri;
mso-hansi-font-family:Calibri;mso-bidi-font-family:Calibri'><span
style='mso-list:Ignore'>-<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span></span><![endif]>The pop-up window related to the advanced find
view currently don’t require mandatory interaction. This should be implemented
to avoid potential issues. Pop ups include:<o:p></o:p></p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:72.0pt;mso-add-space:
auto;text-indent:-18.0pt;mso-list:l1 level2 lfo1'><![if !supportLists]><span
style='font-family:"Courier New";mso-fareast-font-family:"Courier New"'><span
style='mso-list:Ignore'>o<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;
</span></span></span><![endif]>IconSelector<o:p></o:p></p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:72.0pt;mso-add-space:
auto;text-indent:-18.0pt;mso-list:l1 level2 lfo1'><![if !supportLists]><span
style='font-family:"Courier New";mso-fareast-font-family:"Courier New"'><span
style='mso-list:Ignore'>o<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;
</span></span></span><![endif]>DateTimeSelector<o:p></o:p></p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:72.0pt;mso-add-space:
auto;text-indent:-18.0pt;mso-list:l1 level2 lfo1'><![if !supportLists]><span
style='font-family:"Courier New";mso-fareast-font-family:"Courier New"'><span
style='mso-list:Ignore'>o<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;
</span></span></span><![endif]>UseListInputWindow<o:p></o:p></p>

If the find view is already open when constellation opens, its minimum size is not restricted. This is not a problem if the find view is opened after the application is running.

The FindViewStateIOProvider class does not currently cater for the AdvancedFindView. This isn’t a major issue. It just means if there are specific searches that users want to save for when they come back, they won’t be there.


<p class=MsoNormal><b>AutoSaveStartUp.java<o:p></o:p></b></p>


IMPORTANT I have commented out the implementation of auto save for the sake of the CI tests passing. Before this   is ever merged in uncomment that code or look into the CI tests failing due   to auto save triggering. The file is "autosavestartup.java" and the   commented out lines are 58-120.

I would assume this could be added under experimental for the time being to get some user feedback on it once any relevant notes above have been addressed. The current way to open it is CTRL + B (just so it didn't overwrite the existing find view).

antares1470 commented:
I've managed to resolve the autosave issue so that should no longer be a blocker from merging this in

antares1470 reviewed (approved):
Thanks for all your effort on this @Atlas139mkm
